File: cv1.pl
% Author: Martin Pilat
% Date: 3. 2. 2014
% Last Updated: 19. 2. 2020

%%%%%%%%%%%%%%
% Prolog - uvod a program na rodinne vztahy
%%%%%%%%%%%%%%

% Program v prologu je v zasade databaze faktu a vztahu mezi nimi napsana pomoci 
% Hornovskych klauzuli. Kazda klauzule je napsana ve tvaru `H:-B`., kde `H` je 
% hlava klauzule a `B` je jeji telo. Muzete si to predstavovat i tak, ze `:-` 
% je sipka doleva (tedy implikace zprava doleva), a presne takovy je take 
% vyznam klauzule. Klauzule mohou mit i prazdne telo. Potom se pise jen `H`. 
% Napriklad nasledujici 4 klauzule rikaji, ze (postupne), `honza`, `adam`, 
% `david` a `jiri` jsou muzi.


muz(honza). 
muz(adam).
muz(david).
muz(jiri).

% Stejnym zpusobem dalsi klauzule rikaji, ze `jitka`, `jana` a `lida` jsou zeny.

zena(jitka).
zena(jana).
zena(lida).

% Klauzule mohou mit i vice nez jeden parametr. Nasledujici klauzule vyjadruje, 
% ze `honza` je potomek `adam`a. Podobne pro dalsi klauzule.

%potomek(X, Y)
%X je potomek Y
potomek(honza, adam).
potomek(adam, david).
potomek(jiri, adam).

potomek(jitka, adam).
potomek(jitka, jana).

% Pravidla, ktera plati mezi jednotlivymi klauzulemi, se daji napsat pomoci 
% klauzuli s telem. Napriklad fakt, ze kdyz je `X` potomek `Y`, a zaroven je 
% `X` zena, tak `X` je dcera `Y`. V logice bychom to mohli zapsat jako `zena(X) 
% & potomek(X,Y) -> dcera(X,Y)`. V Prologu napiseme to same, jen pomoci 
% prologovske syntaxe, tedy:

dcera(X,Y):-zena(X),potomek(X,Y).
syn(X,Y):-muz(X),potomek(X,Y).

% Vsimnete si, ze jmena konstant zacinaji malym pismenem a jmena promennych 
% velkym. Podobne jako jsme prave nadefinovali `dcera` a `syn`, muzeme definovat
% i dalsi rodinne vztahy.

deda(X,Y):-potomek(Y,Z),potomek(Z,X),muz(X).
bratr(X,Y):-potomek(X,Z),potomek(Y,Z),X \= Y,muz(X).

% K cemu je zapis `X\=Y` v definici `bratr`? `X\=Y` znamena, ze `X` a `Y` nejsou
% unifikovatelne. Bez tohoto atomu, bychom na dotaz `bratr(jiri, X)`, dostali i 
% odpoved, ze `jiri` je bratr sam sebe. A tim se dostavame k dotazum a k tomu, 
% co se Prologem da vlastne delat. 

% __Ukol__: Zkuste si pohrat s priklady v tomto textu, pridejte si dalsi lidi, 
% nadefinujte si napriklad predikaty `teta`, `babicka`, atd.

% Tento soubor si muzete i stahnout jako platny kod v Prologu. Pustte si 
% interpret Prologu, napr. [SWI-Prolog](http://www.swi-prolog.org/) a zkuste 
% tento soubor nacist (`consult(cv1).` -- musite byt ve spravnem adresari, pokud
%  v nem nejste, pouzijte bud `chdir('cesta/s/lomitky/dopredu')`, nebo 
% `working_directory(_,'cesta/s/lomitky/dopredu/')`. Funguje i zkratka `[cv1].` 
% a pokud mate priponu `.pl` asociovanou s Prologem, staci na soubor normalne \
% kliknout.

% Nyni uz muzete pokladat dotazy. 

% Dotazy se pisi primo v interpretu prologu, za prompt `?-` (odpovedi Prologu 
% jsou potom na samostatnych radkach). Nejjednodussi dotaz je napriklad:

%    ?- muz(jiri). 
%    true. 

% Prolog odpovida `true`, tzn. ze byl schopen najit v databazi (nebo z ni 
% odvodit), ze dotaz je pravdivy. Podobne, kdyz polozite dotaz:

%    ?- dcera(jitka, jana).
%    true.

% Prolog zase odpovi `true`, protoze byl z databaze schopen odvodit, ze 
% `jitka` je dcera `jany`.

% Muzete ale pokladat i slozitejsi dotazy, ktere obsahuji promenne, ty v Prologu
% musi zacinat velkym pismenem:

%    ?- dcera(X, jana). 
%    X = jitka ;
%    false.

% V tomto pripade vam Prolog odpovi, ze `X` je Jana, a ceka, co budete delat. 
% Kdyz zmacknete strednik, pokusi se najit dalsi `X`, ktere odpovida dotazu, 
% kdyz se mu to nepovede, odpovi `false`.

% Dotaz nemusi obsahovat vubec zadne konstanty. V takovem pripade Prolog najde 
% vsechna mozna ohodnoceni promennych v dotazu:

%    ?- syn(X,Y).
%    X = honza,
%    Y = adam ;
%    X = adam,
%    Y = david ;
%    X = jiri,
%    Y = adam.

File: cv1.hs
-- Author: Martin Pilat
-- Date: 3. 4. 2017
-- Updated: 31. 3. 2020

----------------------------------------------------------------------------
--          1. cviceni -- Haskell - typy, zakladni funkce
----------------------------------------------------------------------------

-- Po cvicenich s Prologem se dnes presuneme k dalsimu neproceduralnimu jazyku. 
-- Tentokrat nepujde o logicke programovani, ale o programovani funkcionalni, 
-- ktere si ukazeme na jazyce Haskell.

-- Zakladni vlastnosti funkcionalniho programovani je, ze cely program se sklada
-- z funkci, jejichz vystupy zavisi pouze na jejich vstupech (a ne treba na 
-- stavu programu). Neexistuji zadne globalni promenne. Cely program je nakonec
-- jedna (mnohokrat slozena a pripadne i hodne slozita) funkce.

-- Dalsi vlastnosti funkcionalniho programovani je, ze se v nem casto pracuje s
-- funkcemi, ktere jako parametry maji dalsi funkce.

-- Podivejme se napred, jak v Haskellu pracovat se zakladnimi datovymi typy. 
-- Haskell defaultne pocita s dlouhymi cisly (`Integer`), takze kdyz na vstupu 
-- interpretu napisete `5^87`, on vam bez problemu odpovi
-- `6462348535570528709932880406796584793482907116413116455078125`.

-- Haskell samozrejme zna i dalsi obvykle datove typy: `Int` (pro bezna, 
-- "kratka" cisla), `Bool`, `Float`, `Double`, `Char`, ... (vsechny s obvyklym 
-- vyznamem). Haskell podporuje i usporadane `n`-tice, psane v kulatych 
-- zavorkach. `(Int, Double)` je dvojice, kde prvni prvek je `Int` a druhy 
-- `Double`. Konecne, Haskell take zna seznamy, ktere jste si jiste moc oblibili
-- v Prologu :). Seznam prvku typu `Integer` se zapise jako `[Integer]`, syntaxe
-- pro rozdeleni seznamu na hlavu a telo je `x:xs`. Retezce jsou v Haskellu 
-- seznamy znaku `[Char]`.

-- Zkusme si ted v Haskellu napsat prvni jednoduchou funkci, napriklad vypocet 
-- faktorialu, ukazeme si na ni zakladni syntaxi a take to, jak se k funkcim 
-- pisou typy.

fact :: Integer -> Integer  -- typ
fact n  | n < 1     = error "To teda ne"    -- pro zaporna cisla neni definovan
        | n == 1    = 1                     -- faktorial 0 je 1
        | otherwise = n * fact (n-1)        -- faktorial n je n*faktorial n-1

-- Prvni radka explicitne rika, jakeho typu nase funkce `fact` je. Ze zobrazuje 
-- `Integer` na `Integer`. V interpretu si muzeme jeji typ nechat vypsat pomoci
-- `:t fact` a dostaneme odpoved, ktera presne odpovida prvni radce. Typy neni
-- nutne uvadet, Haskell si je umi odvodit sam, ale je to vhodne, pomaha to pri 
-- ladeni a zaroven to pomuze komukoliv, kdo cte vas kod. Zaroven, zrovna v 
-- pripade faktorialu, si Haskell odvodi typ moc obecne a nechal by vas ho 
-- pocitat i pro necela cisla (ale stejne byste minuli ten konec rekurze a pak
-- by vypsal tu chybovou hlasku).

-- Syntaxi s `|` a `=` se rika straz. Mezi `|` a `=` muzete napsat libovolnou 
-- podminku, je to podobne tomu, kdyz v matematice definujete funkci po castech.

-- Zkusme si ted napsat nekolik zakladnich funkci pro praci se seznamy (vsechny
-- z nich jsou v Haskellu definovane, ale nam se pro procviceni hodi). Pro 
-- prehlednost (a zabraneni kolizim s existujicimi jmeny) budeme pred jejich 
-- standardni nazvy psat `my_`.

-- Prvni uzitecnou funkci je funkce `map`, ktera jako argument vezme funkci a 
-- seznam a aplikuje ji na kazdy prvek seznamu, vrati seznam s vysledky.

my_map :: (a -> b) -> [a] -> [b]
my_map f []     = []
my_map f (x:xs) = (f x):(my_map f xs)

-- Muzeme ji hned vyzkouset v interpretu.

-- >my_map (*2) [1,2,3,4]
-- [2,4,6,8]

-- Vsimnete si, ze funkce muze byt definovana i na vice radkach. Haskell pouziva 
-- expression matching a pouzije definici, ktera je jako prvni a odpovida 
-- zvolenemu tvaru vstupu. V tomto pripade se pro prazdny seznam zavola prvni 
-- radek a kdykoliv jindy radek druhy.

-- Hodi se i par vysvetleni k typu. Operator `(*)` je v Haskellu typu 
-- (zjednodusene) `Int -> Int -> Int`, tzn. ze na vstupu ocekava dve cisla a 
-- vraci cislo. Tim, ze napiseme `(*2)` jsme jedno cislo dodali, a tim dostavame 
-- funkci typu `Int -> Int` (tento zapis funguje pro vsechny operatory a binarni
-- funkce pokud je zapiseme v inline zapisu - ukazeme si pozdeji). Samotny zapis
-- typu funkce `my_map` rika, ze jako prvni argument dostava funkci typu 
-- `(a -> b)`, potom seznam prvku typu `a` a vraci seznam prvku typu `b`. 
-- Vsimnete si, ze obecne napsane typy se pisou malym pismenem.

-- A jeste jedna poznamka, pokud pisete `2*3`, staci psat `*`, pokud ale chcete 
-- pouzit operator `*` jako parametr funkce, musite psat `(*)`. Interpretu
-- muzete tedy napriklad napsat i `(*) 2 3` a on vam odpovi `6` (`* 2 3` ale 
-- nefunguje). Stejne pravidlo samozrejme plati i pro ostatni operatory.

-- Uzitecna funkce je i `take n xs`, ta vezme ze seznamu `xs` prvnich `n` prvku
-- a vrati je.

my_take :: Int -> [a] -> [a]
my_take 0 _                     = []
my_take _ []                    = []
my_take n (x:xs) | n < 0        = error "n must be > 0"
                 | otherwise    = x:(my_take (n-1) xs) 

-- Vsimnete si, ze opet muzeme pouzit `_` jako nazev promenne, ktera nas 
-- nezajima.

-- Vyzkousejme si nasi funkci:

-- > my_take 3 [1,2,3,4,5]
-- [1,2,3]

-- Haskell dovoluje definovat seznamy i o neco obecneji. Muzete definovat rozsah 
-- napriklad jako `[3..10]` je seznam vsech cisel od `3` do `10` (vcetne). 
-- Dokonce horni mez seznamu muze chybet, potom se jedna o nekonecny seznam. 
-- `[1..]` tedy definuje seznam `[1,2,3,4,5,6,7,...]`. Zajimave je, ze i s 
-- takovymi seznamy jde v Haskellu bez problemu pracovat.

-- > my_take 5 [1..]
-- [1,2,3,4,5]

-- V tomhle pripade se projevuje jedna z vlastnosti Haskellu. A tou je line 
-- vyhodnocovani, tzn. ze Haskell nevyhodnocuje vyrazy, dokud nemusi. A presne 
-- to se stalo s nasim nekonecnym seznamem, nikdy jsme nepotrebovali prvky od 6
-- dal, takze Haskell na ne nikdy ani nesahl. Dokonce muzeme zkombinovat nase 
-- dve funkce a napsat funkci, ktera nam vrati seznam prvnich `n` mocnin `2`. 
-- Opet, dalsi mocniny se nepocitaji, kdyz nejsou potreba.

mocninyDvojky :: Int -> [Integer]
mocninyDvojky n = my_take n (my_map (2^) [1..])

-- Dokonce konec teto funkce definuje seznam mocnin dvojek, kdyby se nam treba
-- nekdy hodil.

seznamMocnin = my_map (2^) [1..]

-- Uzitecna je take varianta funkce `take`, a sice `takeWhile`, ktera misto
-- poctu ma funkci, ktera vraci `Bool`. `takeWhile` potom bere prvky ze seznamu
-- dokud tato funkce vraci True.

my_takeWhile :: (a -> Bool)->[a]->[a]
my_takeWhile _ []                 = []
my_takeWhile f (x:xs) | f x       = x:(takeWhile f xs)
                      | otherwise = []

-- Dalsi uzitecnou standardni funkci je `zip`. Ta vytvari ze dvou seznamu jeden 
-- seznam dvojic. Delka vysledneho seznamu odpovida delce kratsiho z obou
-- seznamu.

my_zip :: [a] -> [b] -> [(a,b)] 
my_zip [] _          = []
my_zip _ []          = []
my_zip (x:xs) (y:ys) = (x,y):my_zip xs ys

-- > my_zip [1..5] seznamMocnin
-- [(1,2),(2,4),(3,8),(4,16),(5,32),(6,64),(7,128),(8,256),(9,512),(10,1024)]

-- `zipWith` je take uzitecna, vezme dva seznamy, aplikuje na ne binarni funkci
-- a vysledky ulozi do vysledneho seznamu.

my_zipWith :: (a->b->c) -> [a] -> [b] -> [c]
my_zipWith _ _ []          = []
my_zipWith _ [] _          = []
my_zipWith f (x:xs) (y:ys) = (f x y):my_zipWith f xs ys

-- Pomoci ni se da treba napsat scitani dvou seznamu po slozkach

sectiSeznamy x y = my_zipWith (+) x y

-- Ukazme jeste jednu standardni funkci, `foldl` aplikuje operaci na seznam tak,
-- ze ji vlozi mezi vsechny prvky seznamu, jeste je mozne specifikovat pocatecni
-- hodnotu. Napr. `foldl (+) 7 [1,2,3,4]` spocita `(((7 + 1) + 2) + 3) + 4`.
-- Jeji varianta `foldr` dela totez, ale prvky seznamu bere odzadu (tj. pocita
-- `1 + (2 + (3 + (4 + 7)))`).

my_foldl :: (a->b->a)->a->[b]->a
my_foldl _ init []      = init
my_foldl f init (x:xs)  = my_foldl f (f init x) xs

-- Jeste existuje varianta `foldl1`, ktera jako pocatecni hodnotu vezme prvni
-- prvek seznamu.

my_foldl1 :: (a->a->a)->[a]->a
my_foldl1 f (x:xs) = my_foldl f x xs

-- Ted muzeme treba naprogramovat skalarni soucin dvou seznamu.

skalarniSoucin :: Num a => [a] -> [a] -> a
skalarniSoucin x y = my_foldl1 (+) (my_zipWith (*) x y)

-- Specifikace `Num a` v typu funkce rika, ze `a` musi byt nejaky numericky typ.
-- Jedna se o tzv. typovou tridu, jen si dejte pozor na to, ze typova trida neni
-- sama o sobe typ. 

-- Zkusme jeste neco trochu slozitejsiho (alespon ve smyslu typu vysledne
-- funkce). Funkce `filter` ma jako jeden z parametru funkci, ktera vraci
-- `Bool`, a jako druhy parametr seznam. Vybere ze seznamu prvky, pro ktere je
-- dana funkce `True`.

my_filter :: (a->Bool)->[a]->[a]
my_filter _ []              = []
my_filter f (x:xs)  | f x       = x:(filter f xs)
                    | otherwise = filter f xs

-- Opet mame k dispozici par jednoduchych funkci, ktere `Bool` vraceji, napr. 
-- porovnani `(<,>,<=,>=, ==, /=)`. Kdyz jim opet jeden parametr "obsadime",
-- dostavame funkci, ktera ma jeden parametr a vraci `Bool`. Muzeme tedy napr.
-- ze seznamu cisel vyfiltrovat ta, ktera jsou mensi nez 5.

-- > my_filter (<5) [1,2,3,4,5,43,2,3,4]
-- [1,2,3,4,2,3,4]

-- Pomoci teto funkce ale muzeme udelat mnohem zajimavejsi veci, napr. napsat 
-- quicksort.

qsort :: Ord a => [a] -> [a]
qsort []        = []
qsort (x:xs)    = (qsort mensi)++[x]++(qsort vetsi)
                where 
                    mensi = my_filter (<x) xs
                    vetsi = my_filter (>=x) xs

-- Konstrukce `where` dovoluje definovat lokalne funkce primo uvnitr jine
-- funkce, muze obcas trochu zprehlednit kod (kdybychom ale misto
-- `mensi`/`vetsi` napsali primo tu jejich definici, dostali bychom stejny
-- vysledek). Operator `(++)` spojuje dva seznamy (v case linearnim v delce
-- prvniho z nich). 

-- Haskell pouziva (relativne volnou) 2D syntaxi, tj. zalezi na odsazeni.
-- Podstatne ale je jen za klicovymi slovy `where`, `let`, `do` a `of`. Pravidlo
-- pritom je takove, ze hlubsi odsazeni za temito slovy vytvari novy blok,
-- stejne odsazeni pokracuje v existujicim bloku a mensi odsazeni blok ukoncuje.
-- Pokud by nekoho zajimaly detaily, daji se najit v 
-- [dokumentaci](https://www.haskell.org/onlinereport/lexemes.html#lexemes-layout). 
-- Misto odsazeni lze pouzivat i slozene zavorky a stredniky, v Haskellu to ale 
-- neni moc obvykle.